eth/downloader/downloader.go

```diff
@@ -1153,10 +1153,10 @@ func (d *Downloader) processFullSyncContent(peerHeight uint64) error {
 			if err := d.importBlockResults(results); err != nil {
 				return err
 			}
-		}
-
-		if d.core.CurrentBlock().NumberU64() == peerHeight {
-			return nil
+			// If all the blocks are fetched, we exit the sync process
+			if results[0].Header.NumberU64() == peerHeight {
+				return errNoFetchesPending
```

Comment:
Why is this change necessary? If everything completed without error, why do we want to return an error?

Reply:
if you look at the spawnSync logic, anytime the processes return a nil error, it is ignored. So previously when I was returning nil on the sync complete and we would not exit the downloading process. That means you need to return some form of error or exit the downloader even when nil is returned. I chose the special error case and I am catching it in the downloader (the other change in the commit) so that we don't drop the peer. It follows a similar logic to ErrCancelled.

Reply:
https://github.com/dominant-strategies/go-quai/issues/553